pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,353 @@
+}
+
+func (r *ReconcileNSTemplateSet) ensureNamespace(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, tcNamespace *toolchainv1alpha1.Namespace, userNamespace *corev1.Namespace) error {
+	username := nsTmplSet.GetName()
+	nsName := ToNamespaceName(username, tcNamespace.Type)
+
+	log.Info("provisioning namespace", "namespace", tcNamespace)
+	if err := r.setStatusNamespaceProvisioning(nsTmplSet, nsName); err != nil {
+		return err
+	}
+
+	params := map[string]string{"USER_NAME": username}
+
+	if userNamespace == nil {
+		return r.ensureNamespaceResources(logger, nsTmplSet, tcNamespace, params)
+	}
+	return r.ensureOtherResources(logger, nsTmplSet, tcNamespace, params)
+}
+
+func (r *ReconcileNSTemplateSet) ensureNamespaceResources(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, tcNamespace *toolchainv1alpha1.Namespace, params map[string]string) error {
+	username := nsTmplSet.GetName()
+	nsName := ToNamespaceName(username, tcNamespace.Type)
+
+	tmplContent, err := templates.GetTemplateContent(fmt.Sprintf("%s-%s", nsTmplSet.Spec.TierName, tcNamespace.Type))
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	tmplProcessor := template.NewProcessor(r.client, r.scheme)
+	objs, err := tmplProcessor.Process(tmplContent, params)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	objs = template.Filter(objs, template.RetainNamespaces)
+	for _, rawObj := range objs {
+		obj := rawObj.Object
+		if nsObj, ok := obj.(*unstructured.Unstructured); ok {
+			// set labels
+			labels := nsObj.GetLabels()
+			if labels == nil {
+				labels = make(map[string]string)
+			}
+			labels["owner"] = username
+			nsObj.SetLabels(labels)
+
+			// set owner ref
+			if err := controllerutil.SetControllerReference(nsTmplSet, nsObj, r.scheme); err != nil {
+				return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+					"failed to set controller reference for namespace '%s'", nsName)
+			}
+		}
+	}
+
+	err = tmplProcessor.Apply(objs)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	log.Info("namespace provisioned", "namespace", tcNamespace)
+	if err := r.setStatusProvisioning(nsTmplSet); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ReconcileNSTemplateSet) ensureOtherResources(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, tcNamespace *toolchainv1alpha1.Namespace, params map[string]string) error {
+	username := nsTmplSet.GetName()
+	nsName := ToNamespaceName(username, tcNamespace.Type)
+
+	tmplContent, err := templates.GetTemplateContent(fmt.Sprintf("%s-%s", nsTmplSet.Spec.TierName, tcNamespace.Type))
```

Comment:
This is very similar to what is written in the above function `ensureNamespaceResources`, https://github.com/codeready-toolchain/member-operator/pull/53/files#diff-cffea82f7fdc7427b8c9f0d0fdd55aedR161-R172. Can we create a single function and reuse it in both the places

Reply:
Follow here https://github.com/codeready-toolchain/member-operator/pull/53#issuecomment-532054813